### ISSUE TITLE
Can't use checker-framework with Java 7 based application

### ISSUE THREAD

**Issue by croesch** [typetools/checker-framework#1166]
I have an application that compiles against jdk7 and I would like to use the checker-framework. [The manual](https://checkerframework.org/manual/#installation) says

> Requirement: You must have JDK 7 or later installed. You can get JDK 7 from Oracle or elsewhere.

Adding this to a Java 7 based maven project leads to an error

```
  <profiles>
    <profile>
      <id>checker</id>
      <dependencies>
        <dependency>
          <groupId>org.checkerframework</groupId>
          <artifactId>checker</artifactId>
          <version>2.1.9</version>
          <scope>provided</scope>
        </dependency>
        <dependency>
          <groupId>org.checkerframework</groupId>
          <artifactId>jdk7</artifactId>
          <version>2.1.9</version>
          <scope>provided</scope>
        </dependency>
      </dependencies>
      <build>
        <plugins>
          <plugin>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-dependency-plugin</artifactId>
            <executions>
              <execution>
                <goals>
                  <goal>properties</goal>
                </goals>
              </execution>
            </executions>
          </plugin>
          <plugin>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-compiler-plugin</artifactId>
            <configuration>
              <annotationProcessors>
                <annotationProcessor>org.checkerframework.checker.nullness.NullnessChecker</annotationProcessor>
              </annotationProcessors>
              <compilerArgs>
                <arg>-Xbootclasspath/p:${org.checkerframework:jdk7:jar}</arg>
              </compilerArgs>
            </configuration>
          </plugin>
        </plugins>
      </build>
    </profile>
  </profiles>
```

the error is

```
Caused by: java.lang.NoSuchMethodError: com.sun.source.util.JavacTask.instance(Ljavax/annotation/processing/ProcessingEnvironment;)Lcom/sun/source/util/JavacTask;
	at org.checkerframework.javacutil.AbstractTypeProcessor.init(AbstractTypeProcessor.java:99)
	at org.checkerframework.framework.source.SourceChecker.init(SourceChecker.java:427)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$ProcessorState.<init>(JavacProcessingEnvironment.java:517)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$DiscoveredProcessors$ProcessorStateIterator.next(JavacProcessingEnvironment.java:614)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:707)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1700(JavacProcessingEnvironment.java:97)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1029)
	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1163)
	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1108)
	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:824)
	at com.sun.tools.javac.main.Main.compile(Main.java:439)
```

Looking at [JavaDoc of JavacTask](http://docs.oracle.com/javase/7/docs/jdk/api/javac/tree/com/sun/source/util/JavacTask.html) shows that indeed the method is missing in Java 7.

Which leads to my question: Is the framework intended to be used with Java 7 javac and if so am I doing it wrong or is this a bug?

**Comment by smillst** [comment]
Yes, you can you Java 7.  As installation step 3 says:

>Configure your IDE, build system, or command shell to include the Checker Framework on the classpath. Choose the appropriate section of Chapter 30 for javac (Section 30.1), Ant (Section 30.2), Maven (Section 30.3), Gradle (Section 30.4), IntelliJ IDEA (Section 30.6), Eclipse (Section 30.7), or tIDE (Section 30.8).

Since you are using Maven, see [Section 30.3](https://checkerframework.org/manual/#maven). Please let us know if you have questions about those instructions.

**Comment by croesch** [comment]
Thank you for your quick reply. But these are the steps I made. And it leads to the given `NoSuchMethodError`. Can you reproduce that?

**Comment by smillst** [comment]
Since you are not using Java 8, you need to use the type annotations compiler.  Step 3 of the Maven instructions show how to do this:


```
 <plugin>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>3.3</version>
        <configuration>
            <!-- Change source and target to 1.7 if using Java 7 -->
            <source>1.8</source>
            <target>1.8</target>
            <fork>true</fork>
            <annotationProcessors>
                <!-- Add all the checkers you want to enable here -->
                <annotationProcessor>org.checkerframework.checker.nullness.NullnessChecker</annotationProcessor>
            </annotationProcessors>
            <compilerArgs>
                <!-- location of the annotated JDK, which comes from a Maven dependency -->
                <arg>-Xbootclasspath/p:${annotatedJdk}</arg>
                <!-- Uncomment the following line to use the type annotations compiler. -->
                <!-- <arg>-J-Xbootclasspath/p:${typeAnnotationsJavac}</arg> -->
            </compilerArgs>
        </configuration>
    </plugin>
```

In particular this needs to be uncommented:

```
  <!-- Uncomment the following line to use the type annotations compiler. -->
                <!-- <arg>-J-Xbootclasspath/p:${typeAnnotationsJavac}</arg> -->
```

**Comment by croesch** [comment]
Thank you for your quick answer and help!

I didn't have the fork-flag and together with the typeAnnotations compiler it worked. Wasn't clear enough to me. Sorry for not getting it from the manual ;-)

**Comment by smillst** [comment]
Glad you got it to work.  Do you have any suggestions on how the manual might be improved so this is clearer?

**Comment by croesch** [comment]
I did not read the full manual as it is to heavy for me (for a first try of the framework).

I read 1.3, 30.3 and 26.2.1 (and maybe some little spots and read parts of the [Nullness Checker tutorial](https://github.com/glts/safer-spring-petclinic/wiki))

1. I would at least add a (external site) to the link. As it wasn't clear to me that this may be written by somebody else. And that the information - more specific about installation - should not be used from the nullchecker tutorial as they're completely out of date.

2. ```<!-- The type annotations compiler - uncomment if desired -->```
This suggests (together with 26.2.1  Annotations in comments) that I may decide to not use the annotations compiler cause I only want to use annotations allowed in Java 7 and not use annotations in comments. So I would change this to:
```<!-- The type annotations compiler - uncomment if Java 7 or if desired -->```

3. Maybe you could add a comment why it only works with ```<fork>true</fork>``` and that it is thus mandatory.

**Comment by mernst** [comment]
Christian, thanks for the suggestions!  I appreciate this constructive criticism.

I implemented your suggestion 1 in commit c711599.
I implemented your suggestion 2 in commit b89daeb.
